File: src/simulation_agent.py
from impostor import Impostor
from room import Room
from skeld import Skeld
import numpy as np

class SimulationAgent:
    def __init__(self, a_init_energy, crewmates, a_init_sabotageables):
        self.impostor = Impostor(a_init_energy, a_init_sabotageables)
        self.skeld = Skeld(crewmates)

    def simulate(self):
        step = 0
        while step < 1000:
            perception = self.skeld.get_perception()
            self.impostor.see(perception)
            action = self.impostor.select_action()
            if action == None:
                break
            action.update(self.skeld, self.impostor)
            if action.consume_step():
                self.skeld.update()
            step += 1
            if self.impostor.succeded() or self.impostor.failed():
                break

        if self.impostor.succeded():
            print("EL AGENTE GANOOO")
        else:
            print("EL AGENTE PERDIO SADGE PEPOSAD")

def main():
    initial_energy = np.random.randint(30, 151)
    initial_crewmates = np.random.randint(3, 6)
    sab_rooms = {Room("electrical"), Room("reactor"), Room("weapons")}
    sim_agent = SimulationAgent(initial_energy, initial_crewmates, sab_rooms)
    sim_agent.simulate()


if __name__ == "__main__":
    main()

File: SearchAgent/SearchAgent/SimulationAgent.cpp
#include "SimulationAgent.h"
#include "Action.h"
#include <cassert>
#include <chrono>
#include <iostream>

SimulationAgent::SimulationAgent(int initialEnergy, int initialCrewmates, std::set<int> sabotages, Strategy* strategy) {
    this->perceptionAmmount = 0;
    this->energyUsed = 0;
    this->execTime = std::vector<long long>();
    this->skeld = Skeld(initialCrewmates);
    this->impostor = Impostor(skeld.getAgentRoom(), initialEnergy, this->skeld.getCrewmates(), sabotages, strategy);
}

void SimulationAgent::simulate() {
    
    while (this->perceptionAmmount < 10'000) {
        std::cout << "\n+++ STEP #" << this->perceptionAmmount << " +++\n";
        this->perceptionAmmount++;
        std::vector<int> perception = this->skeld.getPerception();
        this->impostor.see(perception);
        this->skeld.printState();
        this->impostor.printState();
        auto start = std::chrono::high_resolution_clock::now();
        Action* action = this->impostor.selectAction();
        auto end = std::chrono::high_resolution_clock::now();
        if (action == nullptr) {
            assert(0);
            break;
        }
        action->printType();
        action->update(this->skeld, this->impostor);
        this->skeld.updateCrewmates();
        this->energyUsed += action->usesEnergy();
        this->execTime.push_back(std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count());
        if (this->impostor.succeeded() or this->impostor.failed())
            break;
        delete action;
    }
    this->skeld.printState();
    this->impostor.printState();
    if (this->impostor.succeeded())
        std::cout << "El agente gano" << std::endl;
    else if (this->impostor.failed())
        std::cout << "El agente perdio" << std::endl;
    else
        std::cout << "Ni gano ni perdio" << std::endl;
    this->printMetrics();
}

void SimulationAgent::printMetrics() {
    std::cout << "\n--- METRICAS ---\n";
    std::cout << "Cantidad de ciclos de percepciones: " << this->perceptionAmmount << '\n';
    std::cout << "Cantidad de energia gastada: " << this->energyUsed << '\n';
    std::cout << "Tiempos de ejecucion:";
    double avg = 0;
    for (const long long& t : this->execTime) {
        avg += t;
        std::cout << ' ' << t;
    }
    std::cout << "\nTiempo promedio de ejecucion: " << avg / this->execTime.size() << "\n --- ---\n";
}